KEYCLOAK-18502 CIBA, support for additional parameters from the backchannel authentication request and backchannel authentication callback.

leandrobortoli opened the pull request:
draft for the [discussion](https://groups.google.com/u/1/g/keycloak-dev/c/-KSDVpvz9kQ) on the dev mail list.

thomasdarimont commented:
The idea makes sense to me :)

@leandrobortoli Would you mind creating a JIRA issue for this and prefix the commit / pr messages with the JIRA issue number?

leandrobortoli commented:
@thomasdarimont Sure, done. If there is anything missing, let me know.

mposolda commented:
@leandrobortoli Thanks for the PR!

I don't see any issue with the approach you taken and moving creation of UserSession to callback endpoint works fine for me. Moreover, if you have a chance to add the callback SPI as discussed on the keycloak-dev thread https://groups.google.com/g/keycloak-dev/c/-KSDVpvz9kQ/m/dn6kTVXbBAAJ , it will be nice.

@tnorimat @pedroigor Adding you as reviewers of this PR. Does the approach from this PR works for you?

tnorimat commented:
Hello @mposolda , @leandrobortoli ,

I would like to review this PR and add comments.

tnorimat reviewed (commented):
@leandrobortoli This PR's codes themselves are good to me.

I have one question.

If we create User Session and Authenticated Client Session on callback endpoint, how we should treat that situation that `auth_req_id` is expired?

IMO, there are two options.
1. remove its corresponding User Session and Authenticated Client Session.
2. do noting. User Session and Authenticated Client Session will expire and are removed.

My vote is (1) because these session are not used any longer.

@mposolda , @pedroigor  What do you think about it?

tnorimat commented:
@leandrobortoli  I agree with your proposal for introducing new `CIBACallbackEndpointSPI`. It would be better if you include it onto this PR. However, IMO, it might be OK to implement it by other PR.

leandrobortoli commented:
@tnorimat @mposolda I like the approach of adding `CIBACallbackEndpointSPI` on this PR, I haven't work on it yet since I didn't know if you guys would agree on it, so I will need a few days to develop it.

Also regarding the expiration of the `auth_req_id`, for me, it makes sense to remove the session, if everyone agrees I will also add this change to the PR.

mposolda commented:
@leandrobortoli @tnorimat Thanks for the review and the discussion started on it. I am fine with include `CIBACallbackEndpointSPI` in this PR or in the follow-up PR. It might be more clear to add it in the follow-up PR. Let me know whatever you prefer.

Regarding the session creation, there is maybe also an option that session will be still created in the token endpoint as today. But we will use `OAuth2DeviceCodeModel.additionalParams` for track the additional parameters. Currently it is always empty map for CIBA, but I think that additional parameters can be added to it similarly like AuthorizationEndpoint adds additional parameters to the AuthenticationSession notes. And then similarly like AuthenticationSession notes are propagated as userSession notes during userSession creation, the parameters from OAuth2DeviceCodeModel can be propagated as userSession notes for CIBA. But ATM I am not sure if it is better approach than earlier creation of userSession like you did in this PR. What do you think guys?

pedroigor commented:
The main reasons for not creating sessions beforehand but when making a token request are:

* It is aligned with other granted types
* It makes sure sessions are not created when the client gives up to obtain tokens

Looks like @mposolda suggestion might work? Correct me if I'm wrong, but the main problem is forward additional parameters from the AD so that they can be used when running mappers, right?

leandrobortoli commented:
@pedroigor Indeed looks like the suggestion from @mposolda works. I'm also not sure which is the best approach, that "fake" sessionId seems sketchy, but if you guys agree I can make the change to use the `OAuth2DeviceCodeModel.additionalParams`

tnorimat commented:
Hello @leandrobortoli , @pedroigor , @mposolda ,

I agree with @mposolda 's proposal if it can do what @leandrobortoli want to do in this PR.

IMO, the advantage of the current approach of creating a session is that we need not remove sessions when corresponding `auth_req_id` expires.

As for `CIBACallbackEndpointSPI` , I think it might be preferable to implement its default provider that works the same as the current CIBA does.

However, CIBA support is treated as technology preview so that it might be allowed to change the default provider's interface from the current one.

mposolda commented:
Cool, Thanks everyone! So just to summary, I think that we are in the agreement to:
- Leverage `OAuth2DeviceCodeModel.additionalParams` for additional parameters
-  Have `CIBACallbackEndpointSPI` with the default provider, which would work same way as current CIBA callback endpoint logic does

pedroigor commented:
Is the callback SPI necessary if we are already solving the problem with the additional params? Sorry, if I missed other requirements that justify the SPI.

leandrobortoli commented:
Well, the main motivation for the SPI was the additional parameters. But since the communication with the AD it's not standardized by CIBA, and the request is already an SPI, IMO the callback should also be customizable. Also, we are analyzing another requirement we have to return additional parameters generated by the authorization server in the backchannel authorization response. the [draft](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html) states in section 3 that:
 `This specification allows for some extensibility - profiles can be created that define additional parameters to be sent to, and received from, the Backchannel Authentication Endpoint.`
 So, I think keycloak should have the option to retrieve additional parameters not only from the backchannel callback but also from the backchannel authentication request and pass additional parameters in the backchannelEndpointResponse. I'm still going to analyse the best way to do this.

pedroigor commented:
@leandrobortoli Indeed, if one needs to customize the response from the callback endpoint to pass additional data to ADs then we need something else. Not sure though about completely replace the callback endpoint through a custom provider but, perhaps, leverage `AuthenticationChannelProvider` to also allow handling responses (today it only deals with requests).

mposolda commented:
@pedroigor @leandrobortoli yes, leverage `AuthenticationChannelProvider` can be also an option instead of introducing separate SPI. Maybe adding additional method to `AuthenticationChannelProvider` for handle callbacks works just fine? But in any case, I agree that some better for customization for callback response will be nice since it is not standardized in CIBA specification.

leandrobortoli committed:
KEYCLOAK-18502 - Support for additional parameters from the backchannel authentication request and backchannel authentication callback.

leandrobortoli commented:
@tnorimat @mposolda @pedroigor
As discussed, I've changed the PR to add the support for additional parameters on the callback response, those are added on the session as notes with the prefix `ciba_callback_response_additional_param_`.
Also, with the purpose of incrementing the extensibility of the CIBA flow, I've also added the support to additional parameters on the backchannel authentication request, those are forwarded to the AD request and also added to the session as notes with the prefix `ciba_backchannel_request_param_`, this will give developers the option to receive additional params from the client and use it on the AD or in custom protocol mappers. Since the specification states that additional parameters on the request are permitted.

As for the customization of the callback with an SPI or using `AuthenticationChannelProvider`, IMO it would be better in a follow-up PR and focus this PR on the parameters support.

Also, I've changed the issue name, description, and PR Name.

mposolda commented on server-spi-private/src/main/java/org/keycloak/models/OAuth2DeviceTokenStoreProvider.java:
```diff
@@ -72,6 +74,8 @@
      */
     boolean approve(RealmModel realm, String userCode, String userSessionId);
 
+    boolean approve(RealmModel realm, String userCode, String userSessionId, Map<String, String> additionalParams);
```
For the simplification purpose, I vote for keep this new method and remove the existing method `boolean approve(RealmModel realm, String userCode, String userSessionId)` from this provider. The existing invocation of the method `approve(realm, userCode, userSessionId)` from the DeviceGrantType can be changed to include null as the last argument. What do you think?

mposolda reviewed (changes requested):
@leandrobortoli Thanks for the update! I've added one suggestion comment, but otherwise looks great to me!

mposolda commented:
@leandrobortoli And +1 for SPI changes in the follow-up PR

tnorimat commented:
@leandrobortoli I also agree with introducing new SPI by follow-up PR.

mposolda commented:
@leandrobortoli Do you please have a chance to rebase on latest master and slightly adjust your PR to the stuff from latest master? There are some conflicts with the already merged PR from @tnorimat . I hope this won't be a major rewrite. Sorry I did not spot this earlier...

leandrobortoli commented:
@mposolda Done. Also made the requested change

mposolda reviewed (changes requested):
@leandrobortoli Thanks a lot! The PR looks great to me.

There is just one small issue that there is one additional conflict in the CIBA test due the other PR merged yesterday https://github.com/keycloak/keycloak/commit/9dc7300178a43adc6b5975e6f597f77212da7893 . I am sorry for that, it is not very common that single PR needs to be rebased on top of master two times due the conflicts. I hope fixing the CIBATest conflict won't be hard. I guess it would be mostly about align with the new way of calling "assertThat" and other with the usage of "MatcherAssert" class.

leandrobortoli commented:
@mposolda Don't worry, indeed It was just the new way of calling assertThat. Fixed.

mposolda reviewed (approved):
@leandrobortoli Thanks!
